qubes-gui/src/lib.rs

```diff
@@ -1,61 +1,279 @@
+
+/// Maximum size of a shared memory segment, in bytes
+pub const MAX_WINDOW_MEM: u32 = MAX_WINDOW_WIDTH * MAX_WINDOW_HEIGHT * (DUMMY_DRV_FB_BPP / 8);
 
-/// Message types
-#[non_exhaustive]
-#[repr(u32)]
-pub enum Msg {
-    /// A key has been pressed
-    KeyPress = 124,
-    /// A button has been pressed
-    Button = 125,
-    /// Crossing
-    Crossing,
-    /// Daemon ⇒ agent: A window has just acquired focus.
-    Focus,
-    /// Daemon ⇒ agent, obsolete.
-    Resize,
-    /// Agent ⇒ daemon: Creates a window.
-    Create,
-    /// Agent ⇒ daemon: Destroys a window.
-    Destroy,
-    /// Bidirectional: Map a window.
-    Map,
-    /// Agent ⇒ daemon: Unmap a window
-    Unmap,
-    /// Bidirectional: Configure a window
-    Configure,
-    /// Ask dom0 (only!) to map the given amount of memory into composition
-    /// buffer.  Deprecated.
-    MfnDump,
-    /// Agent ⇒ daemon: Redraw given area of screen.
-    ShmImage,
-    /// Daemon ⇒ agent: Request that a window be destroyed.
-    Close,
-    /// Daemon ⇒ agent, deprecated, DO NOT USE
-    Execute,
-    /// Daemon ⇒ agent: Request clipboard data.
-    ClipboardReq,
-    /// Agent ⇒ daemon: Reply with clipboard data.
-    ClipboardData,
-    /// Agent ⇒ daemon: Set the title of a window.  Called MSG_WMNAME in C.
-    SetTitle,
-    /// Daemon ⇒ agent: Update the keymap
-    KeymapNotify,
-    /// Agent ⇒ daemon: Dock a window
-    Dock,
-    /// Agent ⇒ daemon: Set window manager hints.
-    WindowHints,
-    /// Agent ⇒ daemon: Set window manager flags.
-    WindowFlags,
-    /// Agent ⇒ daemon: Set window class.
-    WindowClass,
-    /// Agent ⇒ daemon: Send shared memory dump
-    WindowDump,
-    /// Agent ⇒ daemon: Set cursor type
-    Cursor,
+/// Number of bytes in a shared page
+pub const XC_PAGE_SIZE: u32 = 1 << 12;
+
+/// Maximum permissable number of shared memory pages in a single segment using
+/// deprecated privcmd-based shared memory
+pub const MAX_MFN_COUNT: u32 = (MAX_WINDOW_MEM + XC_PAGE_SIZE - 1) >> 12;
+
+/// Maximum permissable number of shared memory pages in a single segment using
+/// grant tables
+pub const MAX_GRANT_REFS_COUNT: u32 = (MAX_WINDOW_MEM + XC_PAGE_SIZE - 1) >> 12;
+
+/// GUI agent listening port
+pub const LISTENING_PORT: i16 = 6000;
+
+/// Type of grant refs dump messages
+pub const WINDOW_DUMP_TYPE_GRANT_REFS: u32 = 0;
+
+macro_rules! enum_const {
+    (
+        #[repr($t: ident)]
+        $(#[$i: meta])*
+        $p: vis enum $n: ident {
+            $(
+                $(#[$j: meta])*
+                ($const_name: ident, $variant_name: ident) $(= $e: expr)?
+            ),*$(,)?
+        }
+    ) => {
+        $(#[$i])*
+        #[repr($t)]
+        #[non_exhaustive]
+        $p enum $n {
+            $(
+                $(#[$j])*
+                $variant_name $(= $e)?,
+            )*
+        }
+
+        $(
+            $(#[$j])*
+            $p const $const_name: $t = $n::$variant_name as $t;
+        )*
+    }
+}
```

Comment:
This is a very unusual thing to have a need for, and I would expect a detailed comment explaining the rationale. In particular, it's more usual to parse an integer into an enum at the boundary, and then actually use the enum type.

(one way to do this is with the `num_enum::TryFromPrimitive` derive.